src/Elm/Parser/Patterns.elm

```diff
@@ -8,38 +8,57 @@ import Elm.Parser.State exposing (State)
 import Elm.Parser.Tokens as Tokens
 import Elm.Syntax.ModuleName exposing (ModuleName)
 import Elm.Syntax.Node as Node exposing (Node)
-import Elm.Syntax.Pattern exposing (Pattern(..), QualifiedNameRef)
+import Elm.Syntax.Pattern as Pattern exposing (Pattern(..), QualifiedNameRef)
 import Parser as Core exposing ((|=))
 
 
-tryToCompose : Node Pattern -> Parser State (Node Pattern)
-tryToCompose x =
+composedWith : Parser State PatternComposedWith
+composedWith =
     Combine.maybeIgnore Layout.layout
         |> Combine.continueWith
             (Combine.oneOf
-                [ Core.keyword "as"
+                [ Tokens.asToken
                     |> Combine.fromCoreIgnore Layout.layout
                     |> Combine.continueWithCore
-                        (Node.parserCoreMap (\y -> Node.combine AsPattern x y)
+                        (Node.parserCoreMap PatternComposedWithAs
```

Comment:
I wonder whether putting a function (instead of a custom type+pattern matching) would not have been better for performance (similar to the "apply" trick)
```elm
Node.parserCoreMap (\x -> \y -> Node.combine AsPattern x y)
-- Maybe it should be \y -> \x ->
-- I'm not sure yet
```

Reply:
I see that later you use the custom type version (https://github.com/stil4m/elm-syntax/pull/233/commits/090a985201d801fdc55d67b5529162cb82ed6033). That ends up being faster I'm guessing?

Reply:
I did benchmark this. And if the intermediate union type is relatively cheap to construct/reuse it's clearly faster.

But I might be blind for the more heavy intermediates because I do vastly prefer this style over "apply style".
Tho I have also pretty extensively compared these changes using the 100 run benchmark and not found a consistent difference between the styles, so :shrug:

Reply:
Yeah I was thinking that creating a new data structure would be slower than a function, but maybe creating a function (with a lot of closure?) is slower?
But alright, I think the new style is pretty nice :+1:
I loved seeing the journey of the API evolution :smile: